File: android/app/src/main/java/com/treforged/forged/widgets/WidgetSnapshot.java
package com.treforged.forged.widgets;

import android.content.Context;
import android.content.SharedPreferences;

import org.json.JSONException;
import org.json.JSONObject;

public class WidgetSnapshot {
    static final String PREFS_NAME = "forgenta_widget";
    static final String KEY = "snapshot";

    public final double monthEndCash;
    public final double netWorth;
    public final String currency;
    public final long updatedAtMs; // epoch millis

    public WidgetSnapshot(double monthEndCash, double netWorth, String currency, long updatedAtMs) {
        this.monthEndCash = monthEndCash;
        this.netWorth = netWorth;
        this.currency = currency;
        this.updatedAtMs = updatedAtMs;
    }

    /**
     * After this long without an update, the stored figure stops being
     * information. A widget shows a number without anyone opening the app, so
     * NOBODY OPENS THE APP TO CHECK WHAT THEIR HOME SCREEN ALREADY TOLD THEM —
     * which makes a stale number worse than a blank one. A blank prompts a tap.
     * Seven days is sized against what these figures are: month-end cash and net
     * worth move with every transaction, and a week is long enough to contain a
     * paycheck and a rent payment.
     */
    static final long STALE_AFTER_MS = 7L * 24 * 60 * 60 * 1000;

    /**
     * A snapshot, or null when there is not one worth drawing.
     *
     * ⚠️ THIS USED TO DEFAULT MISSING NUMBERS TO ZERO. `optDouble("netWorth", 0)`
     * meant a malformed or partial snapshot rendered as "$0" in confident gold,
     * and a user whose net worth genuinely is zero looked identical to one whose
     * data never arrived. That is the same error as a gauge drawing a value it
     * never read: there is no failure state a person can see. Missing fields now
     * return null, and the provider draws "--" instead.
     */
    public static WidgetSnapshot load(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        return parse(prefs.getString(KEY, null));
    }

    /**
     * The trust rules, split out from {@link #load} so they can be EXERCISED.
     *
     * `load` needs an Android Context and therefore only runs on a device; this
     * takes the raw string and runs on the JVM, so `./gradlew testDebugUnitTest`
     * presses every rule below. Splitting it is the difference between a
     * read-and-reason change and a tested one.
     *
     * @return a snapshot, or null when there is not one worth drawing.
     */
    public static WidgetSnapshot parse(String json) {
        if (json == null) return null;
        try {
            JSONObject obj = new JSONObject(json);
            // has() before read: absent is NOT zero.
            if (!obj.has("monthEndCash") || !obj.has("netWorth") || !obj.has("updatedAtMs")) {
                return null;
            }
            double cash = obj.getDouble("monthEndCash");
            double worth = obj.getDouble("netWorth");
            // NaN and Infinity are what a division by a missing denominator looks
            // like, and NumberFormat will happily print them.
            if (Double.isNaN(cash) || Double.isInfinite(cash)) return null;
            if (Double.isNaN(worth) || Double.isInfinite(worth)) return null;

            long updatedAt = obj.getLong("updatedAtMs");
            if (updatedAt <= 0) return null;

            return new WidgetSnapshot(cash, worth, obj.optString("currency", "USD"), updatedAt);
        } catch (JSONException e) {
            return null;
        }
    }

    /** True once this snapshot is too old to be shown as a figure. */
    public boolean isStale(long nowMs) {
        return nowMs - this.updatedAtMs > STALE_AFTER_MS;
    }

    public static void save(Context context, double monthEndCash, double netWorth, String currency) {
        try {
            JSONObject obj = new JSONObject();
            obj.put("monthEndCash", monthEndCash);
            obj.put("netWorth", netWorth);
            obj.put("currency", currency);
            obj.put("updatedAtMs", System.currentTimeMillis());
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                .edit()
                .putString(KEY, obj.toString())
                .apply();
        } catch (JSONException e) {
            // The write failed, so the widget keeps whatever it had. `load` will
            // refuse to draw it once it is older than STALE_AFTER_MS, which is
            // the backstop that makes this ignore safe.
        }
    }
}

File: src/lib/widget-snapshot.ts
/**
 * Whether the app has a number good enough to put on someone's HOME SCREEN.
 *
 * A widget is the least forgiving surface this app has. It shows a figure without
 * anyone opening anything, which means **nobody opens the app to check what the
 * home screen already told them**. A stale or defaulted value there is worse than
 * a blank one: a blank prompts a tap, a wrong number ends the conversation.
 *
 * So the rule from every other surface applies hardest here — never render a
 * figure you did not actually read — and it is enforced in two places on purpose:
 *
 *   - HERE, on the way out: the app refuses to SEND a number it does not have.
 *   - In `WidgetSnapshot.java`, on the way in: the widget refuses to RENDER a
 *     snapshot that is absent, malformed, or too old.
 *
 * Either one alone leaves a hole. A widget installed on a phone that has not
 * opened the app in a month has a perfectly well-formed snapshot; only the
 * reader can catch that one.
 */

/**
 * After this long without an update, the number on the home screen stops being
 * information and starts being a claim about a past that may not hold. Seven
 * days is chosen against what the figures actually are: month-end cash and net
 * worth move with every transaction, and a week is long enough that a balance
 * can have changed by a paycheck and a rent payment.
 */
export const WIDGET_STALE_AFTER_MS = 7 * 24 * 60 * 60 * 1000;

export interface WidgetPayload {
  monthEndCash: number;
  netWorth: number;
  currency: string;
  updatedAt: string;
}

export interface WidgetInputs {
  monthEndCash: number | null | undefined;
  netWorth: number | null | undefined;
  currency: string | null | undefined;
  /** False in demo, in partner view, or while the figures are still loading. */
  enabled: boolean;
}

/**
 * The payload to send, or null when the app does not have one worth sending.
 *
 * ⚠️ `null` and `undefined` are NOT zero. A user whose net worth genuinely is
 * zero and a user whose projection has not loaded are different people, and
 * `optDouble(..., 0)` on the Android side used to render both as "$0" in
 * confident gold. Sending nothing leaves the widget saying "open Forgenta to
 * sync", which is true.
 *
 * NaN and Infinity are refused for the same reason: they are what a division by
 * a missing denominator looks like, and `NumberFormat` will happily print "∞".
 */
export function buildWidgetPayload(inputs: WidgetInputs, now: Date): WidgetPayload | null {
  if (!inputs.enabled) return null;

  const cash = inputs.monthEndCash;
  const worth = inputs.netWorth;
  if (typeof cash !== 'number' || !Number.isFinite(cash)) return null;
  if (typeof worth !== 'number' || !Number.isFinite(worth)) return null;

  return {
    monthEndCash: cash,
    netWorth: worth,
    // The user's own currency, not a hardcoded dollar sign. Formatting a
    // non-USD figure with "$" is a wrong number rendered confidently, which is
    // the same failure as a stale one.
    currency: inputs.currency && inputs.currency.trim() !== '' ? inputs.currency : 'USD',
    updatedAt: now.toISOString(),
  };
}

/** True when a snapshot of this age should no longer be shown as a figure. */
export function isSnapshotStale(updatedAt: string | null, now: Date): boolean {
  if (!updatedAt) return true;
  const then = new Date(updatedAt).getTime();
  if (Number.isNaN(then)) return true;
  return now.getTime() - then > WIDGET_STALE_AFTER_MS;
}
